File: wpilibj/src/test/java/edu/wpi/first/wpilibj/AlertTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.networktables.StringArraySubscriber;
import edu.wpi.first.wpilibj.Alert.AlertType;
import edu.wpi.first.wpilibj.simulation.SimHooks;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;
import org.junit.jupiter.api.parallel.ResourceLock;

class AlertTest {
  String m_groupName;

  @BeforeEach
  void setup(TestInfo info) {
    m_groupName = "AlertTest_" + info.getDisplayName();
  }

  @AfterEach
  void checkClean() {
    update();
    assertEquals(0, getActiveAlerts(AlertType.ERROR).length);
    assertEquals(0, getActiveAlerts(AlertType.WARNING).length);
    assertEquals(0, getActiveAlerts(AlertType.INFO).length);
  }

  private String getSubtableName(Alert.AlertType type) {
    switch (type) {
      case ERROR:
        return "errors";
      case WARNING:
        return "warnings";
      case INFO:
        return "infos";
      default:
        return "unknown";
    }
  }

  private StringArraySubscriber getSubscriberForType(Alert.AlertType type) {
    return NetworkTableInstance.getDefault()
        .getStringArrayTopic("/SmartDashboard/" + m_groupName + "/" + getSubtableName(type))
        .subscribe(new String[] {});
  }

  private String[] getActiveAlerts(AlertType type) {
    update();
    try (var sub = getSubscriberForType(type)) {
      return sub.get();
    }
  }

  private void update() {
    SmartDashboard.updateValues();
  }

  private boolean isAlertActive(String text, Alert.AlertType type) {
    return Arrays.asList(getActiveAlerts(type)).contains(text);
  }

  private void assertState(Alert.AlertType type, List<String> state) {
    assertEquals(state, Arrays.asList(getActiveAlerts(type)));
  }

  private Alert makeAlert(String text, Alert.AlertType type) {
    return new Alert(m_groupName, text, type);
  }

  @Test
  void setUnsetSingle() {
    try (var one = makeAlert("one", AlertType.INFO)) {
      assertFalse(isAlertActive("one", AlertType.INFO));
      one.set(true);
      assertTrue(isAlertActive("one", AlertType.INFO));
      one.set(false);
      assertFalse(isAlertActive("one", AlertType.INFO));
    }
  }

  @Test
  void setUnsetMultiple() {
    try (var one = makeAlert("one", AlertType.ERROR);
        var two = makeAlert("two", AlertType.INFO)) {
      assertFalse(isAlertActive("one", AlertType.ERROR));
      assertFalse(isAlertActive("two", AlertType.INFO));
      one.set(true);
      assertTrue(isAlertActive("one", AlertType.ERROR));
      assertFalse(isAlertActive("two", AlertType.INFO));
      one.set(true);
      two.set(true);
      assertTrue(isAlertActive("one", AlertType.ERROR));
      assertTrue(isAlertActive("two", AlertType.INFO));
      one.set(false);
      assertFalse(isAlertActive("one", AlertType.ERROR));
      assertTrue(isAlertActive("two", AlertType.INFO));
    }
  }

  @Test
  void setIsIdempotent() {
    try (var a = makeAlert("A", AlertType.INFO);
        var b = makeAlert("B", AlertType.INFO);
        var c = makeAlert("C", AlertType.INFO)) {
      a.set(true);
      b.set(true);
      c.set(true);

      var startState = List.of(getActiveAlerts(AlertType.INFO));

      b.set(true);
      assertState(AlertType.INFO, startState);

      a.set(true);
      assertState(AlertType.INFO, startState);
    }
  }

  @Test
  void closeUnsetsAlert() {
    try (var alert = makeAlert("alert", AlertType.WARNING)) {
      alert.set(true);
      assertTrue(isAlertActive("alert", AlertType.WARNING));
    }
    assertFalse(isAlertActive("alert", AlertType.WARNING));
  }

  @Test
  void setTextWhileUnset() {
    try (var alert = makeAlert("BEFORE", AlertType.INFO)) {
      assertEquals("BEFORE", alert.getText());
      alert.set(true);
      assertTrue(isAlertActive("BEFORE", AlertType.INFO));
      alert.set(false);
      assertFalse(isAlertActive("BEFORE", AlertType.INFO));
      alert.setText("AFTER");
      assertEquals("AFTER", alert.getText());
      alert.set(true);
      assertFalse(isAlertActive("BEFORE", AlertType.INFO));
      assertTrue(isAlertActive("AFTER", AlertType.INFO));
    }
  }

  @Test
  void setTextWhileSet() {
    try (var alert = makeAlert("BEFORE", AlertType.INFO)) {
      assertEquals("BEFORE", alert.getText());
      alert.set(true);
      assertTrue(isAlertActive("BEFORE", AlertType.INFO));
      alert.setText("AFTER");
      assertEquals("AFTER", alert.getText());
      assertFalse(isAlertActive("BEFORE", AlertType.INFO));
      assertTrue(isAlertActive("AFTER", AlertType.INFO));
    }
  }

  @ResourceLock("timing")
  @Test
  void setTextDoesNotAffectFirstOrderSort() {
    SimHooks.pauseTiming();
    try (var a = makeAlert("A", AlertType.INFO);
        var b = makeAlert("B", AlertType.INFO);
        var c = makeAlert("C", AlertType.INFO)) {
      a.set(true);
      SimHooks.stepTiming(1);
      b.set(true);
      SimHooks.stepTiming(1);
      c.set(true);

      var expectedEndState = new ArrayList<>(List.of(getActiveAlerts(AlertType.INFO)));
      expectedEndState.replaceAll(s -> "B".equals(s) ? "AFTER" : s);
      b.setText("AFTER");

      assertState(AlertType.INFO, expectedEndState);
    } finally {
      SimHooks.resumeTiming();
    }
  }

  @ResourceLock("timing")
  @Test
  void sortOrder() {
    SimHooks.pauseTiming();
    try (var a = makeAlert("A", AlertType.INFO);
        var b = makeAlert("B", AlertType.INFO);
        var c = makeAlert("C", AlertType.INFO)) {
      a.set(true);
      assertState(AlertType.INFO, List.of("A"));
      SimHooks.stepTiming(1);
      b.set(true);
      assertState(AlertType.INFO, List.of("B", "A"));
      SimHooks.stepTiming(1);
      c.set(true);
      assertState(AlertType.INFO, List.of("C", "B", "A"));

      SimHooks.stepTiming(1);
      c.set(false);
      assertState(AlertType.INFO, List.of("B", "A"));

      SimHooks.stepTiming(1);
      c.set(true);
      assertState(AlertType.INFO, List.of("C", "B", "A"));

      SimHooks.stepTiming(1);
      a.set(false);
      assertState(AlertType.INFO, List.of("C", "B"));

      SimHooks.stepTiming(1);
      b.set(false);
      assertState(AlertType.INFO, List.of("C"));

      SimHooks.stepTiming(1);
      b.set(true);
      assertState(AlertType.INFO, List.of("B", "C"));

      SimHooks.stepTiming(1);
      a.set(true);
      assertState(AlertType.INFO, List.of("A", "B", "C"));
    } finally {
      SimHooks.resumeTiming();
    }
  }
}

File: wpilibc/src/test/native/cpp/AlertTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <algorithm>
#include <chrono>
#include <map>
#include <string>
#include <utility>
#include <vector>

#include <fmt/format.h>
#include <gtest/gtest.h>
#include <networktables/NetworkTableInstance.h>
#include <networktables/StringArrayTopic.h>

#include "frc/Alert.h"
#include "frc/simulation/SimHooks.h"
#include "frc/smartdashboard/SmartDashboard.h"

using namespace frc;
using enum Alert::AlertType;
class AlertsTest : public ::testing::Test {
 public:
  ~AlertsTest() {
    // test all destructors
    Update();
    EXPECT_EQ(GetSubscriberForType(ERROR).Get().size(), 0ul);
    EXPECT_EQ(GetSubscriberForType(WARNING).Get().size(), 0ul);
    EXPECT_EQ(GetSubscriberForType(INFO).Get().size(), 0ul);
  }

  std::string GetGroupName() {
    const ::testing::TestInfo* testInfo =
        ::testing::UnitTest::GetInstance()->current_test_info();
    return fmt::format("{}_{}", testInfo->test_suite_name(), testInfo->name());
  }

  template <typename... Args>
  Alert MakeAlert(Args&&... args) {
    return Alert(GetGroupName(), std::forward<Args>(args)...);
  }

  std::vector<std::string> GetActiveAlerts(Alert::AlertType type) {
    Update();
    return GetSubscriberForType(type).Get();
  }

  bool IsAlertActive(std::string_view text, Alert::AlertType type) {
    auto activeAlerts = GetActiveAlerts(type);
    return std::find(activeAlerts.begin(), activeAlerts.end(), text) !=
           activeAlerts.end();
  }

  void Update() { frc::SmartDashboard::UpdateValues(); }

 private:
  std::string GetSubtableName(Alert::AlertType type) {
    switch (type) {
      case ERROR:
        return "errors";
      case WARNING:
        return "warnings";
      case INFO:
        return "infos";
      default:
        return "unknown";
    }
  }

  const nt::StringArraySubscriber GetSubscriberForType(Alert::AlertType type) {
    return nt::NetworkTableInstance::GetDefault()
        .GetStringArrayTopic(fmt::format("/SmartDashboard/{}/{}",
                                         GetGroupName(), GetSubtableName(type)))
        .Subscribe({});
  }
};

#define EXPECT_STATE(type, ...) \
  EXPECT_EQ(GetActiveAlerts(type), (std::vector<std::string>{__VA_ARGS__}))

TEST_F(AlertsTest, SetUnsetSingle) {
  auto one = MakeAlert("one", INFO);
  EXPECT_FALSE(IsAlertActive("one", INFO));
  one.Set(true);
  EXPECT_TRUE(IsAlertActive("one", INFO));
  one.Set(false);
  EXPECT_FALSE(IsAlertActive("one", INFO));
}

TEST_F(AlertsTest, SetUnsetMultiple) {
  auto one = MakeAlert("one", ERROR);
  auto two = MakeAlert("two", INFO);
  EXPECT_FALSE(IsAlertActive("one", ERROR));
  EXPECT_FALSE(IsAlertActive("two", INFO));
  one.Set(true);
  EXPECT_TRUE(IsAlertActive("one", ERROR));
  EXPECT_FALSE(IsAlertActive("two", INFO));
  one.Set(true);
  two.Set(true);
  EXPECT_TRUE(IsAlertActive("one", ERROR));
  EXPECT_TRUE(IsAlertActive("two", INFO));
  one.Set(false);
  EXPECT_FALSE(IsAlertActive("one", ERROR));
  EXPECT_TRUE(IsAlertActive("two", INFO));
}

TEST_F(AlertsTest, SetIsIdempotent) {
  auto a = MakeAlert("A", INFO);
  auto b = MakeAlert("B", INFO);
  auto c = MakeAlert("C", INFO);
  a.Set(true);

  b.Set(true);
  c.Set(true);

  const auto startState = GetActiveAlerts(INFO);

  b.Set(true);
  EXPECT_STATE(INFO, startState);

  a.Set(true);
  EXPECT_STATE(INFO, startState);
}

TEST_F(AlertsTest, DestructorUnsetsAlert) {
  {
    auto alert = MakeAlert("alert", WARNING);
    alert.Set(true);
    EXPECT_TRUE(IsAlertActive("alert", WARNING));
  }
  EXPECT_FALSE(IsAlertActive("alert", WARNING));
}

TEST_F(AlertsTest, SetTextWhileUnset) {
  auto alert = MakeAlert("BEFORE", INFO);
  EXPECT_EQ("BEFORE", alert.GetText());
  alert.Set(true);
  EXPECT_TRUE(IsAlertActive("BEFORE", INFO));
  alert.Set(false);
  EXPECT_FALSE(IsAlertActive("BEFORE", INFO));
  alert.SetText("AFTER");
  EXPECT_EQ("AFTER", alert.GetText());
  alert.Set(true);
  EXPECT_FALSE(IsAlertActive("BEFORE", INFO));
  EXPECT_TRUE(IsAlertActive("AFTER", INFO));
}

TEST_F(AlertsTest, SetTextWhileSet) {
  auto alert = MakeAlert("BEFORE", INFO);
  EXPECT_EQ("BEFORE", alert.GetText());
  alert.Set(true);
  EXPECT_TRUE(IsAlertActive("BEFORE", INFO));
  alert.SetText("AFTER");
  EXPECT_EQ("AFTER", alert.GetText());
  EXPECT_FALSE(IsAlertActive("BEFORE", INFO));
  EXPECT_TRUE(IsAlertActive("AFTER", INFO));
}

TEST_F(AlertsTest, SetTextDoesNotAffectFirstOrderSort) {
  frc::sim::PauseTiming();

  auto a = MakeAlert("A", ERROR);
  auto b = MakeAlert("B", ERROR);
  auto c = MakeAlert("C", ERROR);

  a.Set(true);
  frc::sim::StepTiming(1_s);
  b.Set(true);
  frc::sim::StepTiming(1_s);
  c.Set(true);

  auto expectedEndState = GetActiveAlerts(ERROR);
  std::replace(expectedEndState.begin(), expectedEndState.end(),
               std::string("B"), std::string("AFTER"));
  b.SetText("AFTER");

  EXPECT_STATE(ERROR, expectedEndState);
  frc::sim::ResumeTiming();
}

TEST_F(AlertsTest, MoveAssign) {
  auto outer = MakeAlert("outer", INFO);
  outer.Set(true);
  EXPECT_TRUE(IsAlertActive("outer", INFO));

  {
    auto inner = MakeAlert("inner", WARNING);
    inner.Set(true);
    EXPECT_TRUE(IsAlertActive("inner", WARNING));
    outer = std::move(inner);
    // Assignment target should be unset and invalidated as part of move, before
    // destruction
    EXPECT_FALSE(IsAlertActive("outer", INFO));
  }
  EXPECT_TRUE(IsAlertActive("inner", WARNING));
}

TEST_F(AlertsTest, MoveConstruct) {
  auto a = MakeAlert("A", INFO);
  a.Set(true);
  EXPECT_TRUE(IsAlertActive("A", INFO));
  Alert b{std::move(a)};
  EXPECT_TRUE(IsAlertActive("A", INFO));
  b.Set(false);
  EXPECT_FALSE(IsAlertActive("A", INFO));
  b.Set(true);
  EXPECT_TRUE(IsAlertActive("A", INFO));
}

TEST_F(AlertsTest, SortOrder) {
  frc::sim::PauseTiming();
  auto a = MakeAlert("A", INFO);
  auto b = MakeAlert("B", INFO);
  auto c = MakeAlert("C", INFO);
  a.Set(true);
  EXPECT_STATE(INFO, "A");
  frc::sim::StepTiming(1_s);
  b.Set(true);
  EXPECT_STATE(INFO, "B", "A");
  frc::sim::StepTiming(1_s);
  c.Set(true);
  EXPECT_STATE(INFO, "C", "B", "A");

  frc::sim::StepTiming(1_s);
  c.Set(false);
  EXPECT_STATE(INFO, "B", "A");

  frc::sim::StepTiming(1_s);
  c.Set(true);
  EXPECT_STATE(INFO, "C", "B", "A");

  frc::sim::StepTiming(1_s);
  a.Set(false);
  EXPECT_STATE(INFO, "C", "B");

  frc::sim::StepTiming(1_s);
  b.Set(false);
  EXPECT_STATE(INFO, "C");

  frc::sim::StepTiming(1_s);
  b.Set(true);
  EXPECT_STATE(INFO, "B", "C");

  frc::sim::StepTiming(1_s);
  a.Set(true);
  EXPECT_STATE(INFO, "A", "B", "C");

  frc::sim::ResumeTiming();
}
